### PR TITLE
feat: report unused Java private members

### LINKED ISSUE #164: Add Java detection for unused private members using the existing unused-class-member guide
Java could reuse the existing `unused-class-member` smell to report unused private fields and methods detected by PMD.

The TypeScript plugin already provides a guide for this smell. It asks the agent to check for false positives caused by reflection or frameworks before deciding whether to remove, connect, or extract the reported members.

## Proposed changes

- Enable PMD rules `UnusedPrivateField` and `UnusedPrivateMethod` in the Java ruleset.
- Map both rules to `unused-class-member` in the Java PMD sensor.
- Make the existing guide available to Java, potentially by moving it into the generic plugin.
- Adapt the guide's diagnostic rendering to accommodate the Java sensor's output.

The current guide renders `issue.details.file`, `issue.details.line`, and `issue.details.name`. The integration should ensure that Java diagnostics provide compatible details or adjust the template to support them.

Existing guide:
https://github.com/habit-hooks/habit-hooks/blob/main/plugins/typescript/src/habit_hooks_typescript/guides/unused-class-member.md

PMD rules:
https://docs.pmd-code.org/latest/pmd_rules_java_bestpractices.html#unusedprivatefield
https://docs.pmd-code.org/latest/pmd_rules_java_bestpractices.html#unusedprivatemethod

## Expected behavior

When PMD reports an unused private field or method, Habit Hooks produces `unused-class-member` and makes the corresponding guidance available to the agent.

Detection should not imply automatic deletion. The existing guidance should continue to ask the agent to consider reflection, framework lifecycle hooks, and other uses that static analysis may miss.

## Suggested tests

- An unused private field maps to `unused-class-member`.
- An unused private method maps to the same smell.
- Referenced private members are not reported in representative Java fixtures.
- Annotation and framework exclusions follow the configured PMD behavior.
- The guide resolves and renders correctly with Java diagnostic details.
- Existing TypeScript guide resolution and rendering continue to work.

### PR DESCRIPTION
**Pull request by renanfranca**
## What this changes

This PR adds Java detection for unused private fields and methods. It enables PMD's `UnusedPrivateField` and `UnusedPrivateMethod` rules in the bundled fallback ruleset and maps both results to the existing `unused-class-member` smell.

The Java integration uses the source file as the issue key. The `key` and `details.file` fields identify that file. Each issue also preserves the line where PMD found the problem, the complete diagnostic message, and the distinct source `pmd:UnusedPrivateField` or `pmd:UnusedPrivateMethod`.

When a file contains several violations, they continue to share the snooze behavior for that entire file. PMD remains responsible for deciding whether a member is reachable, applying exclusions and properties, and producing custom messages. A ruleset supplied by the project still replaces the bundled fallback ruleset instead of being combined with it.

This PR also moves `unused-class-member.md` from the TypeScript plugin into `generic`. There is now one shared source of guidance that applies across programming languages. The guide covers fields, methods, properties, reflection, annotations and decorators, serialization, lifecycle hooks, and other uses that static analysis may not detect.

The guide asks readers to investigate the member before changing it. Depending on what they find, they can connect the intended caller, extract a cohesive unit, or remove a member that is confirmed to be unused. The guide does not suggest automatic deletion.

## Why the guide selects `name` or `message`

Knip and PMD provide different data. Knip identifies the member through `details.name`. PMD provides the complete description in its JSON report, but it does not provide a separate field dedicated to the member name.

Extracting a name from PMD text would make custom messages fragile. It would also couple the adapter to wording that PMD does not guarantee as part of its detector contract.

The shared template therefore uses `details.name` when that field is defined. Otherwise, it uses `details.message`. If both fields are present, `name` takes precedence. The mapper does not inspect the programming language, interpret diagnostic text, or add another display field.

Moving the guide does not turn `generic` into a detector, and it does not enable `generic` automatically. When `generic` is omitted, language sensors still report the smell and the existing fallback without coaching remains in effect.

## Documentation and decisions

The smell vocabulary now records both Java mappings, the behavior of the bundled fallback ruleset, and the authority of a ruleset supplied by the project. It also explains that Java issues identify the source file, while TypeScript issues identify the member by name.

The decision record preserves the previous ownership by the TypeScript plugin as historical context and explains why #164 moves the single shared guide to `generic`. The executable guide specifications demonstrate how the displayed value is selected, and the Java plugin specifications demonstrate the behavior of the real PMD integration.

## Validation

- Scenarios using PMD 7.26.0 cover unused and referenced members, exclusions controlled by PMD, project rulesets that omit or enable the rules, and customization through `ignoredFieldNames`.
- The complete test suite passed: `1070 passed, 4 skipped`.
- The focused suite covering Java, the generic guide, Knip, mapper behavior, specifications, and packaging passed: `138 passed`.
- Tests of the installed wheels without access to the source tree passed: `7 passed`. These tests include both Java rules and confirm that only the `generic` plugin packages the shared guide.
- `uv run ruff check --config pyproject.toml src tests plugins` passed.
- `uv run habit-hooks --all` completed with zero findings.
- GitHub CI passed on Ubuntu and Windows.

The enforced severity remains unchanged. Projects that use the bundled Java fallback ruleset may now see unused private members that were not reported before. Projects that supply their own PMD ruleset retain full control, and this change never modifies a reported member automatically.

Closes #164

Any feedback on the contracts, examples, or guidance is very welcome.

### CODE REVIEW

**Comment by renanfranca** [comment]
Hello @devill ,

I would like to let you know that this PR is ready to review when you have the time. 😀

**Comment by devill** [comment]
Thanks! I'll review this tomorrow.

**Comment by devill** [comment]
Sorry for the late response. This PR is a bit harder to review, because it seems to do more than one thing, and also surfaced pre existing issues. Some of the issues are pre existing anti-patterns that got repeated. I'll leave a few comments.

File: docs/guide-includes.spec.md
Comment (devill) [inline docs/guide-includes.spec.md:1]: This file shouldn't have changed at all. It has nothing to do with the java plugin. A single PR should either focus on a plugin or a core functionality. Please revert these changes, or if they are needed open a separate PR for them

File: docs/smell-vocabulary.md
Comment (devill) [inline docs/smell-vocabulary.md:1]: Same issue here. This is a pre existing issue though, so I'll handle it as a separate change, but this file should know nothing about the plugins at all.

File: plugins/generic/src/habit_hooks_generic/guides/unused-class-member.md [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+Once those hidden uses are ruled out, choose the design that makes the type
+honest: connect an intended caller, extract a cohesive group of members into a
+real unit, or remove a member confirmed to be dead together with code that
+exists only to support it. Do not remove a member automatically from this
+report alone.
+
+The unused members:
+
+{% for issue in issues -%}
```
Comment (devill) [inline plugins/generic/src/habit_hooks_generic/guides/unused-class-member.md]: Should be using the line level issues include here. Seems like the plugins sensor isn't faithfully following the interface defined by core, making this custom template necessary. Please make sure the plugin follows the pre defined interface.

File: plugins/generic/tests/test_unused_member_coaching_adds_no_detector.py
Comment (devill) [inline plugins/generic/tests/test_unused_member_coaching_adds_no_detector.py:1]: Unnecessary test.

File: plugins/java/docs/java-plugin.spec.md
Comment (devill) [inline plugins/java/docs/java-plugin.spec.md:1]: spec.md files are meant to be user facing documentation with executable examples. It's not a place for testing every edge case. We use unit tests for that.

**Review by devill — COMMENTED** [review]
Please consider the suggestions.

File: docs/guide-includes.spec.md
Comment (renanfranca) [reply docs/guide-includes.spec.md:1]: Thanks, reverted.

File: docs/smell-vocabulary.md
Comment (renanfranca) [reply docs/smell-vocabulary.md:1]: Thanks, reverted.

File: plugins/generic/src/habit_hooks_generic/guides/unused-class-member.md (hunk 1 above)
Comment (renanfranca) [reply plugins/generic/src/habit_hooks_generic/guides/unused-class-member.md]: Thanks for the guidance. In this pull request, I moved the unused-class-member guide from the TypeScript plugin to the generic plugin so that Java and TypeScript could share it. Following your feedback, the generic guide now uses includes/line_level_issues.md. I also updated both the PMD and Knip sensors to emit details.content, which allows both languages to follow the core interface. Is this the scope you intended? Or would you prefer that I restore the TypeScript guide and Knip sensor, and apply the interface change only to the Java PMD sensor?

File: plugins/generic/tests/test_unused_member_coaching_adds_no_detector.py
Comment (renanfranca) [reply plugins/generic/tests/test_unused_member_coaching_adds_no_detector.py:1]: Thanks, removed.

File: plugins/java/docs/java-plugin.spec.md
Comment (renanfranca) [reply plugins/java/docs/java-plugin.spec.md:1]: Thanks, leave only one test.

**Comment by renanfranca** [comment]
> Please consider the suggestions. 

Thank you. I addressed the suggestions. I asked a question here to better understand one of yours suggestions: https://github.com/habit-hooks/habit-hooks/pull/165#discussion_r4001097898

Thank you for your time 😀

File: plugins/generic/src/habit_hooks_generic/guides/unused-class-member.md (hunk 1 above)
Comment (devill) [reply plugins/generic/src/habit_hooks_generic/guides/unused-class-member.md]: Yes, that is the scope I intended, thanks!

**Comment by devill** [comment]
PR merged, thanks for the contribution! (An sorry for the slow response. I've been taking time off)

File: plugins/generic/src/habit_hooks_generic/guides/unused-class-member.md (hunk 1 above)
Comment (renanfranca) [reply plugins/generic/src/habit_hooks_generic/guides/unused-class-member.md]: > Yes, that is the scope I intended, thanks!

Thank you too :)

**Comment by renanfranca** [comment]
> PR merged, thanks for the contribution! (An sorry for the slow response. I've been taking time off)

You are welcome 😄 ! All good 👍
